### DISCUSSION TITLE
공학인증 요건에 대하여

### DISCUSSION THREAD

**Discussion by gmltn9233** [Sejong-Java-Study/Gimme-Gonghak-Auth#34] [Ideas]
## 1.  AbeekType
현재 공학인증 AbeekType을 아래와 같은 기준으로 나누어 적용하고 있습니다.
* **case1**: 전선, 전공주제, 전필, 전공
* **case2**: 전문교양
* **case3**: MSC
* **case4**: BSM

그러나 현재 [공학인증센터](https://abeek.sejong.ac.kr/abeek/program0302.html) 년도별 조건을 확인해보면 2022,2023년도 에만 `전문교양` 이 아닌 `교양`으로 나와 있습니다. (24년부터는 다시 `전문교양` 으로 바뀜, 모든 학과)
![image](https://github.com/user-attachments/assets/78ede738-99e6-4785-b8a2-b6e52168d73a)
그래서 이를 처리할 방법이 필요하다고 생각합니다. 

제 생각으로서는 아래 2가지 방법이 있을 것 같습니다.
1. `CourseCategoryConst`와 `GonghakCalculateService.stackUserGonghakCredit` 등 **abeekTyp**e이 사용되는 코드에 `교양` 조건 추가
2. 데이터를 `교양` 을 `전문교양` 포맷으로 바꾸어서 저장
 
## 2. 교양과목의 변화
22년 이전에는 들어야 할 교양과목이 정해져있던 반면, 22년도부터 **교양과목**이 `필수 과목`과 `선택과목`으로 나누어졌습니다.
![image](https://github.com/user-attachments/assets/7b1cd935-87f0-4389-b836-1ca82830da17)
하지만 현재 기준으로 삼고있는  **교과과정 표** 에는 필수 교양만 나와있고 선택과목이 나와있지 않아, 추가적으로 작업이 필요할 것 같습니다.
![image](https://github.com/user-attachments/assets/4880620f-b64b-4730-8f16-282d41c0176e)

## 3. 학과 추가
그리고 현재 학과를 `컴퓨터공학과` , `전자정보통신공학과` 만 취급하고 있는데, `데사`,`소웨`,`정보보호` 과 모두 기존과 동일한 **AbeekType(전문교양, BSM,전공,전공설계)** 을 가지고 있어 추가하는데에는 로직상 문제가 없어 보입니다.

하지만 정보보호과는 다른 학과와 달리 BSM 과목이 `인증필수`와 `인증선택`으로 나누어져 현재로서는 제외하는게 맞다고 생각합니다.
![image](https://github.com/user-attachments/assets/6c2db4d8-a987-4216-92d1-86fd5af46a4a)

또한 기준으로 삼고 있는 **교과과정 표**가 학과마다 달라서 기존의 `컴공`이나 `전정통` 과 같은 방식으로 데이터 양식을 통일하는 과정이 필요할 것 같습니다.
소프트웨어 학과의 **교과과정 표**는 아래와 같이 `인필`, `인필(BSM)`,`인필(전공)` 으로 나누고
![image](https://github.com/user-attachments/assets/34e5ac16-0399-4e49-a813-037ee4f7c5ae)
심지어 데이터 사이언스학과는 `BSM` 제외 `인증 필수`, `인증 선택` 으로 나누어집니다.
![image](https://github.com/user-attachments/assets/92454873-e58f-40d3-a98b-54c99d1d4f9d)
그래서 위 2개의 학과를 추가하려면 데이터 형식을 맞추는 수작업이 필요하다고 생각됩니다.

**수작업이 귀찮더라도 출시할때 적어도 4개 학과 정도는 있어야 좋을것 같다는 생각인데, 팀원들의 의견을 묻고 싶습니다.**
이것 말고도 위의 수정사항으로 출시가 늦어질것 같다면 `컴공`,`전전통`만 출시 후 이용량이나 사용자의 요구가 많아지면 추가하는 방식도 있을 것 같습니다.

**Comment by gmltn9233** [comment]
또한 현재 `Status`와 관련된 계산과정, 추천서비스 코드를 읽어보며 주석을 달고있는데 완료되면 PR할테니, 코드 작성자(@jiseop121) 리뷰 부탁드립니다.

**Reply by whxogus215** [reply]
초기 서비스에서 4개 학과 정도 지원이 되면 좋을 것 같다는 의견에 동의합니다!

**Reply by whxogus215** [reply]
1. AbeekType 글에서 24년 부터는 다시 `전문교양`으로 바뀐거 아닌가요? 오타인지 확인 부탁드립니다.

**Reply by gmltn9233** [reply]
> 1. AbeekType 글에서 24년 부터는 다시 `전문교양`으로 바뀐거 아닌가요? 오타인지 확인 부탁드립니다.

오타가 맞습니다 😥 수정했습니다!

**Comment by whxogus215** [comment]
`1. CourseCategoryConst와 GonghakCalculateService.stackUserGonghakCredit 등 abeekType이 사용되는 코드에 교양 조건 추가`
`2. 데이터를 교양을 전문교양 포맷으로 바꾸어서 저장`

이 두 가지 방법 중에서는 1번이 낫다고 생각합니다. 저의 개인적인 생각이지만 데이터의 변경사항에 대해 어플리케이션이 충분히 대응할 수 있는게 중요하다고 생각합니다. 2번의 경우, 어플리케이션 단에서 해결하는 방법은 아닌 것 같다는 생각이 들었습니다.

**Reply by gmltn9233** [reply]
1번 항목(AbeekType)에 대해서는 저도 동의합니다. 그런데 학과 추가에 있어서는 학과마다 데이터의 형식이 상이하여 데이터를 읽어들이는데 학과별 로직을 구현하는 것 보다는 데이터의 포맷을 기존 방식으로 통일하는 것이 효과적이라고 생각이 드는데, 이에 의견을 묻고싶습니다.

**Reply by whxogus215** [reply]
추후 구현 로직을 좀 더 분석해서 정리해보고 의견 드리겠습니다

**Reply by whxogus215** [reply]
혹시 어느 코드들을 중점적으로 보면 좋은지 알려주실 수 있나요 😢

**Reply by gmltn9233** [reply]
> 혹시 어느 코드들을 중점적으로 보면 좋은지 알려주실 수 있나요 😢

현재 간단한 흐름은 아래와 같습니다.
1. `abeek` 테이블에서 `courseCategory` column을  `enum CourseCategoryConst` 으로 읽어 저장합니다.
2. `GonghakCalculateService.getResultRatio` 안의 `stackUserGonghakCredit` 로직에서 **courseCategory** 항목을 기준으로 이수한 학점을 쌓아갑니다.
![image](https://github.com/user-attachments/assets/ffb82f9d-68a0-49df-981e-b2728e09c482)

그런데 현재 abeek 테이블의 포맷이 `컴퓨터 공학과` 의 교과과정표를 기준삼은거라 데이터를 `abeek`테이블에 **import** 방식으로 입력하는 현재 상황으로선 어차피 데이터 포맷팅이 필요합니다.

그래서 1번의  `전문교양`이 `교양`으로 바뀐 문제는 **abeekType** 과 위의 로직에 `교양`을 추가하여 해결하고, 3번의 학과 추가문제는 데이터 포멧을 일치시키는게 좋다는게 제 생각입니다!

**Comment by jiseop121** [comment]
3가지 주제가 다 연관이 있어서 통합적으로 코맨트 남깁니다.
년도/학과별 abeek의 구분법과 교과과정표는 같은 것도 있으나 차이가 심한 경우가 많습니다.
그래서 추후 어떻게 보완될지 몰라 일단 각 핵심로직 클래스에 대해 인터페이스화를 진행했던것이고

그래서 저도 `1. CourseCategoryConst와 GonghakCalculateService.stackUserGonghakCredit 등 abeekType이 사용되는 코드에 교양 조건 추가`
라는 의견에 동의합니다.

일단 정보보호학과나, 다른 년도의 같은 학과든, 너무 예외사항이 많은 학과는 제외하고 abeekType에 타입을 추가하는 방식으로 확장하는 방식이 적절할 것 같습니다.

**Reply by gmltn9233** [reply]
확인했습니다. 교양 선택과목의 경우 4개 학과 모두 아래의 항목으로 동일하므로 6개 과목만 22~24 교과과정표에 수기로 추가하면 될것 같습니다.
* (역사와사상영역) 세계사, 동서양의사상과윤리
* (경제와사회영역) 경제학, 경영학
* (문화와예술영역) 컴퓨터게임과메타버스, 융합예술의이해

그런데 인증필수 항목의 경우 추가 로직이 필요할것 같은데, 22년도 이후로만 로직을 분리하기엔 힘들것 같다는 생각이 듭니다.
이 부분에 대해선 어떻게 생각하시나요?

**Reply by gmltn9233** [reply]
> 확인했습니다. 교양 선택과목의 경우 4개 학과 모두 아래의 항목으로 동일하므로 6개 과목만 22~24 교과과정표에 수기로 추가하면 될것 같습니다.
> 
> * (역사와사상영역) 세계사, 동서양의사상과윤리
> * (경제와사회영역) 경제학, 경영학
> * (문화와예술영역) 컴퓨터게임과메타버스, 융합예술의이해
> 
> 그런데 인증필수 항목의 경우 추가 로직이 필요할것 같은데, 22년도 이후로만 로직을 분리하기엔 힘들것 같다는 생각이 듭니다. 이 부분에 대해선 어떻게 생각하시나요?

개인적인 생각으로서는 현재 전공필수를 모두 이수하였는지 체크하는 로직이 따로 없듯이 필수항목 체크를 따로 진행하지않고 선후수나 필수과목은 유저에게 책임을 맡기는게 좋다고 최선이라고 생각이듭니다..😥

**Reply by jiseop121** [reply]
네 좋은 것 같습니다.
전공이든 새로 생긴 교양이든 필수에 따른 추가적인 서버 로직을 생성하기 보다는, 필수체크는 유저에게 책임을 주되, 안내문을 추가하는 것이 좋아보입니다.

**Comment by gmltn9233** [comment]
그럼 아래와 같이 정리하겠습니다. 확인 댓글 남겨주시면 Discussion 닫겠습니다~👍
## 1. AbeekType
`CourseCategoryConst`와 `GonghakCalculateService.stackUserGonghakCredit` 등 `abeekType`이 사용되는 코드에 `교양` 조건 추가

## 2. 교양과목의 변화 
22~24 년도 `Abeek` 테이블에 아래의 6 과목 수기로 추가
* (역사와사상영역) 세계사, 동서양의사상과윤리
* (경제와사회영역) 경제학, 경영학
* (문화와예술영역) 컴퓨터게임과메타버스, 융합예술의이해

**필수과목과 선택과목은 따로 구분하지 않고 안내문과 같은 방식으로 유저에게 책임을 전가함**

## 3. 학과 추가
기존 `컴퓨터공학과`,`전자정보통신공학과` 이외에 `데이터사이언스학과`,`소프트웨어학과` 2개를 추가하여 총 4개의 학과로 서비스 출시

**단, 교과과정표가 기존 `Abeek` 테이블 포맷 (컴퓨터공학과) 와 다른 `데이터사이언스학과`,`소프트웨어학과` 는 기존 `컴퓨터공학과` 포맷을 기준으로 데이터를 가공해서 사용함**

**Reply by whxogus215** [reply]
확인했습니다 정리 감사합니다~
